synapse/handlers/federation.py

```diff
@@ -748,6 +748,19 @@ async def on_make_join_request(
         # (and return a 404 otherwise)
         room_version = await self.store.get_room_version(room_id)
 
+        if await self.store.is_partial_state_room(room_id):
+            # If our server is still only partially joined, we can't give a complete
+            # response to make_join, so return a 404 as we would if we weren't in the
+            # room at all.
```

Comment:
I'm confused about why we can't respond to `/make_join`s. The subsequent `/send_join` request will fail of course, but surely we can produce a "correct" `/make_join` response?

Is it because we may not know about the previous membership event of the joining user, so we can't cite it in the auth events?

Reply:
To be honest, I just did what I was told :P (perhaps from @richvdh if I remember).

Looking at the code, in order to respond to a `/make_join`, we need:
* to know some `prev_event_ids` for the event (i.e. have reasonably up-to-date forward extremities)
* to be able to calculate auth events for the event we're generating (according to `EventBuilder.build`)
    * This says they will be 'calculated based on the room state at the prev_events'
    * This calls `compute_state_after_events` → `resolve_state_groups_for_events` (with `await_full_state = True` implicitly).
    * Can we avoid this? I'm not sure I have enough confidence to say.

> Is it because we may not know about the previous membership event of the joining user, so we can't cite it in the auth events?

This sounds like a reasonable explanation for at least one case where we can't answer properly.


I also am somewhat tempted to say that failing `/make_join` at least doesn't give any impression that we might be able to answer a `/send_join`. I don't know if any current homeservers are actually happy to `/make_join` and then `/send_join` it on a different server, so to some degree there's no point in entertaining the `/make_join`?

Happy to hear more thoughts.

Reply:
Agreed, if we can't handle /send_join we might as well not entertain the /make_join.

Can we add some words to that effect?